File: app/src/main/java/me/echeung/moemoekyun/adapter/ViewPagerAdapter.java
package me.echeung.moemoekyun.adapter;

import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentPagerAdapter;
import android.util.SparseArray;
import android.view.ViewGroup;

import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.List;

import me.echeung.moemoekyun.R;
import me.echeung.moemoekyun.ui.fragment.SongsFragment;
import me.echeung.moemoekyun.ui.fragment.UserFragment;

public class ViewPagerAdapter extends FragmentPagerAdapter {

    private final SparseArray<WeakReference<Fragment>> fragments = new SparseArray<>();
    private final List<TabHolder> holders = new ArrayList<>();

    @NonNull
    private final Context context;

    public ViewPagerAdapter(@NonNull Context context, FragmentManager fragmentManager) {
        super(fragmentManager);
        this.context = context;

        // Tabs
        add(UserFragment.class, R.drawable.ic_person_white_24dp);
        add(SongsFragment.class, R.drawable.ic_audiotrack_white_24dp);
    }

    @NonNull
    @Override
    public Object instantiateItem(@NonNull ViewGroup container, int position) {
        Fragment fragment = (Fragment) super.instantiateItem(container, position);
        WeakReference<Fragment> fragmentRef = fragments.get(position);
        if (fragmentRef != null) {
            fragmentRef.clear();
        }
        fragments.put(position, new WeakReference<>(fragment));
        return fragment;
    }

    @Override
    public void destroyItem(ViewGroup container, int position, Object object) {
        super.destroyItem(container, position, object);
        WeakReference<Fragment> mWeakFragment = fragments.get(position);
        if (mWeakFragment != null) {
            mWeakFragment.clear();
        }
    }

    @Override
    public int getCount() {
        return holders.size();
    }

    @Override
    public Fragment getItem(int position) {
        TabHolder holder = holders.get(position);
        return Fragment.instantiate(context, holder.className, holder.params);
    }

    @Override
    public CharSequence getPageTitle(int position) {
        return "";
    }

    public int getIcon(int position) {
        return holders.get(position).drawableId;
    }

    private void add(@NonNull Class<? extends Fragment> className, int drawableId) {
        TabHolder holder = new TabHolder();
        holder.className = className.getName();
        holder.drawableId = drawableId;
        holder.params = null;

        holders.add(holder);
        notifyDataSetChanged();
    }

    private final static class TabHolder {
        String className;
        int drawableId;
        Bundle params;
    }

}

File: app/src/main/java/me/echeung/moemoekyun/adapter/ViewPagerAdapter.kt
package me.echeung.moemoekyun.adapter

import android.content.Context
import android.util.SparseArray
import android.view.ViewGroup
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import androidx.fragment.app.FragmentPagerAdapter
import me.echeung.moemoekyun.R
import me.echeung.moemoekyun.ui.fragment.SongsFragment
import me.echeung.moemoekyun.ui.fragment.UserFragment
import java.lang.ref.WeakReference
import java.util.ArrayList

class ViewPagerAdapter(private val context: Context, fragmentManager: FragmentManager) : FragmentPagerAdapter(fragmentManager) {

    private val fragments = SparseArray<WeakReference<Fragment>>()
    private val holders = ArrayList<TabHolder>()

    init {
        // Tabs
        add(UserFragment::class.java, R.drawable.ic_person_24dp)
        add(SongsFragment::class.java, R.drawable.ic_audiotrack_24dp)
    }

    override fun instantiateItem(container: ViewGroup, position: Int): Any {
        val fragment = super.instantiateItem(container, position) as Fragment
        fragments.get(position)?.clear()
        fragments.put(position, WeakReference(fragment))
        return fragment
    }

    override fun destroyItem(container: ViewGroup, position: Int, `object`: Any) {
        super.destroyItem(container, position, `object`)
        val fragment = fragments.get(position)
        fragment?.clear()
    }

    override fun getCount(): Int {
        return holders.size
    }

    override fun getItem(position: Int): Fragment {
        val holder = holders[position]
        return Fragment.instantiate(context, holder.className)
    }

    override fun getPageTitle(position: Int): CharSequence {
        return ""
    }

    fun getIcon(position: Int): Int {
        return holders[position].drawableId
    }

    private fun add(className: Class<out Fragment>, drawableId: Int) {
        holders.add(TabHolder(className.name, drawableId))
        notifyDataSetChanged()
    }

    private data class TabHolder(
        val className: String,
        val drawableId: Int
    )
}
